Artifacts/windows-vsts-build-agent/vsts-agent-install.ps1

```diff
@@ -1,148 +1,293 @@
+    }
+
+    if (![string]::IsNullOrWhiteSpace($WorkDirectory) -and !(Test-ValidPath -Path $WorkDirectory))
+    {
+        Write-Error "Work directory '$WorkDirectory' is not a valid path."
+    }
 }
 
-if ($workDirectory -ne "" -and !(Test-Path -Path $workDirectory -IsValid -ErrorAction Ignore))
+function Test-ValidPath
 {
-    Write-Error "Work Directory '$workDirectory' is not a valid path."
+    param(
+        [string] $Path
+    )
+    
+    $isValid = Test-Path -Path $Path -IsValid -PathType Container
+    
+    try
+    {
+        [IO.Path]::GetFullPath($workDirectory) | Out-Null
+    }
+    catch
+    {
+        $isValid = $false
+    }
+    
+    return $isValid
 }
 
-$currentLocation = Split-Path -parent $MyInvocation.MyCommand.Definition
+function Download-AgentPackage
+{
+    [CmdletBinding()]
+    param(
+        [string] $VstsAccount,
+        [string] $VstsUserPassword
+    )
+    
+    # Create a temporary directory to download from VSTS the agent package (agent.zip) to, and then launch the configuration.
+    $agentTempFolderName = Join-Path $env:temp ([System.IO.Path]::GetRandomFileName())
+    New-Item -ItemType Directory -Force -Path $agentTempFolderName | Out-Null
+
+    $agentPackagePath = "$agentTempFolderName\agent.zip"
+    $serverUrl = "https://$VstsAccount.visualstudio.com"
+    $vstsAgentUrl = $serverUrl + '/_apis/distributedtask/packages/agent/win7-x64?$top=1&api-version=3.0'
+    $vstsUser = "AzureDevTestLabs"
 
-# Create a temporary directory to download from VSTS the agent package (agent.zip) to, and then launch the configuration.
-$agentTempFolderName = Join-Path $env:temp ([System.IO.Path]::GetRandomFileName())
-New-Item -ItemType Directory -Force -Path $agentTempFolderName
+    $retryCount = 3
+    $retries = 1
+    do
+    {
+        try
+        {
+            $basicAuth = ("{0}:{1}" -f $vstsUser, $vstsUserPassword) 
+            $basicAuth = [System.Text.Encoding]::UTF8.GetBytes($basicAuth)
+            $basicAuth = [System.Convert]::ToBase64String($basicAuth)
+            $headers = @{ Authorization = ("Basic {0}" -f $basicAuth) }
 
-$serverUrl = "https://$vstsAccount.visualstudio.com"
-$vstsAgentUrl = $serverUrl + '/_apis/distributedtask/packages/agent/win7-x64?$top=1&api-version=3.0'
-$vstsUser = "AzureDevTestLabs"
+            $agentList = Invoke-RestMethod -Uri $vstsAgentUrl -Headers $headers -Method Get -ContentType application/json
+            $downloadUrl = $agentList.value[0].downloadUrl
+            Invoke-WebRequest -Uri $downloadUrl -Headers $headers -Method Get -OutFile "$agentPackagePath" | Out-Null
+            break
+        }
+        catch
+        {
+            $exceptionText = ($_ | Out-String).Trim()
+            $retries++
+                
+            if ($retries -ge $retryCount)
+            {
+                Write-Error "Failed to download agent due to $exceptionText"
+            }
+                
+            Start-Sleep -Seconds 30 

```

Comment:
The last retry, you will sleep an extra 30 seconds for no reason. (Unless write-error exits??)